src/infra/core/src/pull_service_impl.rs

```diff
@@ -419,20 +422,64 @@ impl PullServiceImpl {
         Ok(results)
     }
 
+    #[async_recursion::async_recursion]
     async fn transform_multi(
         &self,
         batch: &[PullItem], // TODO: Move to avoid cloning
-        listener: Option<Arc<dyn TransformMultiListener>>,
+        transform_listener: Option<Arc<dyn TransformMultiListener>>,
+        compacting_listener: Option<Arc<dyn CompactingMultiListener>>,
+        retry_with_compaction: bool,
     ) -> Result<Vec<PullResponse>, InternalError> {
         let transform_requests = batch.iter().map(|pi| pi.local_ref.clone()).collect();
 
         let transform_results = self
             .transform_svc
-            .transform_multi(transform_requests, listener)
+            .transform_multi(transform_requests, transform_listener.clone())
             .await;
 
         assert_eq!(batch.len(), transform_results.len());
 
+        if retry_with_compaction {
+            let mut compacting_target_refs = vec![];
+            for (dataset_ref, transform_result) in &transform_results {
+                if let Err(transform_err) = transform_result
+                    && let TransformError::InvalidInterval(_) = transform_err
+                {
+                    compacting_target_refs.push(dataset_ref.clone());
+                }
+            }
+
+            if !compacting_target_refs.is_empty() {
+                let compacting_results = self
+                    .compacting_svc
+                    .compact_multi(
+                        compacting_target_refs,
+                        CompactingOptions {
+                            keep_metadata_only: true,
+                            ..Default::default()
+                        },
+                        compacting_listener,
+                    )
+                    .await;
+
+                let retry_items: Vec<_> = std::iter::zip(batch, compacting_results)
```

Comment:
If `transform_multi` is called on multiple datasets and only one of them fails with `InvalidInterval` then `batch` and `compacting_results` will have different lengths and `zip` will not do what we expect it to do here.

This will be the case with dataset graph like:
```
A -> B
C -> D
```

A and C will be pulled as roots first, then `transform_multi` will be called on B and D in one batch, and if A was compacted and D wasn't - you'll get into this case.

Please add a test for the case like this before fixing - it's a tricky one.

Reply:
Such as we removed this logic from pull service to transform service where we work with single dataset it is not a case anymore